zarr/codecs.py

```diff
@@ -485,6 +485,100 @@ def __repr__(self):
 codec_registry[Delta.codec_id] = Delta
 
 
+class AsType(Codec):
+    """Filter to convert data between different types.
+
+    Parameters
+    ----------
+    encode_dtype : dtype
+        Data type used to store the data.
+    decode_dtype : dtype, optional
+        Data type of the data read in.
+
+    Notes
+    -----
+    If `encode_dtype` is of lower precision than `decode_dtype`, please be
+    aware that data loss can occur by writing data to disk using this filter.
+    No checks are made to ensure the casting will work in that direction and
+    data corruption will occur.
+
+    Examples
+    --------
+    >>> import zarr
+    >>> import numpy as np
+    >>> x = np.arange(100, 120, 2, dtype='i1')
+    >>> x
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+    >>> f = zarr.AsType(encode_dtype=x.dtype, decode_dtype='i8')
+    >>> y = f.decode(x)
+    >>> y
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118])
+    >>> z = f.encode(y)
+    >>> z
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+
+    """  # flake8: noqa
+
+    codec_id = 'astype'
+
+    def __init__(self, encode_dtype, decode_dtype):
+        self.encode_dtype = np.dtype(encode_dtype)
+        self.decode_dtype = np.dtype(decode_dtype)
+
+    def encode(self, buf):
+
+        # view input data as 1D array
+        arr = _ndarray_from_buffer(buf, self.decode_dtype)
+
+        # setup encoded output
+        enc = np.empty_like(arr, dtype=self.encode_dtype)
+
+        # Convert and copy
+        enc[...] = arr.astype(self.encode_dtype)
```

Comment:
I think this might mean two copies are being done here? I.e., could you remove a copy if you replaced lines 534 and 537 with just:

```
enc = arr.astype(self.encode_dtype)
```

...?